InfiniTAM/infinitam_ros_node.cpp

```diff
@@ -190,113 +233,286 @@ bool InfinitamNode::startInfinitam(std_srvs::SetBool::Request& request,
   } else {
     cpu_triangles = main_engine_->GetMesh()->triangles;
   }
 
+  // Read the memory and store it in a new array.
   ITMMesh::Triangle* triangleArray = cpu_triangles->GetData(MEMORYDEVICE_CPU);
 
+  ROS_ERROR_COND(main_engine_->GetMesh()->noTotalTriangles < 1,
+                 "The mesh has too few triangles, only: %d",
+                 main_engine_->GetMesh()->noTotalTriangles);
+
+  //  if (save_cloud_to_file_system_) {
+  if (true) {
+    // write a STL or OBJ File to the file system.
+    std::string filename =
+        "../output_" + std::to_string(ros::Time::now().toSec()) + ".stl";
+    main_engine_->GetMesh()->WriteSTL(filename.c_str());
+    std::string filename2 =
+        "../output_" + std::to_string(ros::Time::now().toSec()) + ".obj";
+    main_engine_->GetMesh()->WriteOBJ(filename2.c_str());
+  }
+
+  if (publish_point_cloud_) {
+    // Publish point cloud.
+    pcl::PointCloud<pcl::PointXYZ>::Ptr point_cloud_pcl(
+        new pcl::PointCloud<pcl::PointXYZ>);
+
+    extractITMMeshToPclCloud(triangleArray, point_cloud_pcl);
+    LOG(INFO) << "got point cloud";
+
+    sensor_msgs::PointCloud2 point_cloud_msg;
+    pcl::toROSMsg(*point_cloud_pcl, point_cloud_msg);
+    point_cloud_msg.header.frame_id = camera_frame_id_;
+    point_cloud_msg.header.stamp = ros::Time::now();
+
+    complete_point_cloud_pub_.publish(point_cloud_msg);
+  }
+
+  if (publish_mesh_) {
+    ROS_INFO_STREAM("publish_mesh_");
+
+    // Get the Mesh as PCL PolygonMesh .
+    extractITMMeshToPolygonMesh(triangleArray, mesh_ptr_);
+
+    ROS_INFO_STREAM("Loaded a PolygonMesh with "
+                    << mesh_ptr_->cloud.width * mesh_ptr_->cloud.height
+                    << " points and " << mesh_ptr_->polygons.size()
+                    << " polygons.");
+
+    pcl::io::saveOBJFile("../awesome_mesh.obj", *mesh_ptr_);
+
+    //     Convert PCL PolygonMesh into ROS shape_msgs Mesh.
+    convertPolygonMeshToRosMesh(mesh_ptr_, ros_scene_mesh_ptr_);
+    ROS_INFO_STREAM("Got ROS Mesh.");
+
+    // Publish ROS Mesh.
+    complete_mesh_pub_.publish(*ros_scene_mesh_ptr_);
+    ROS_INFO_STREAM("ROS Mesh published.");
+  }
+
+  if (rm_triangle_from_cuda_memory) {
+    delete cpu_triangles;
+  }
+  ROS_INFO_STREAM("publishMap end!");
+  return true;
+}
+
+void InfinitamNode::extractITMMeshToPolygonMesh(
+    const ITMMesh::Triangle* triangleArray,
+    pcl::PolygonMesh::Ptr polygon_mesh_ptr) {
+  if (triangleArray == NULL) {
+    ROS_ERROR_STREAM("triangleArray == NULL");
+    return;
+  }
+
+  std::size_t nr_triangles = 0;
+  std::size_t nr_points = 0;
+  nr_triangles = main_engine_->GetMesh()->noTotalTriangles;
+  nr_points = nr_triangles * 3;
+  ROS_INFO_STREAM("nr_triangles:  " << nr_triangles);
+  ROS_INFO_STREAM("nr_points:  " << nr_points);
+
+  pcl::PointCloud<pcl::PointXYZ>::Ptr point_cloud_pcl(
+      new pcl::PointCloud<pcl::PointXYZ>);
+
+  point_cloud_pcl->width = nr_points;
+  point_cloud_pcl->height = 1;
+  point_cloud_pcl->is_dense = true;
```

Comment:
Object discovery already does this, should I also do that here and iterate through all points?